test/api/test_workflows.py

```diff
@@ -964,6 +964,31 @@ def test_workflow_request(self):
         self.dataset_populator.wait_for_history(history_id, assert_ok=True)
 
     @skip_without_tool("cat")
+    def test_cancel_workflow_when_history_deleted(self):
+        workflow = self.workflow_populator.load_workflow_from_resource("test_workflow_pause")
+        uploaded_workflow_id = self.workflow_populator.create_workflow(workflow)
+        history_id = self.dataset_populator.new_history()
+        hda1 = self.dataset_populator.new_dataset(history_id, content="1 2 3")
+        index_map = {
+            '0': self._ds_entry(hda1),
+        }
+        invocation_id = self.__invoke_workflow(
+            history_id,
+            uploaded_workflow_id,
+            index_map,
+        )
+
+        # Wait for all the datasets to complete, make sure the workflow invocation
+        # is not complete.
+        invocation = self._invocation_details(uploaded_workflow_id, invocation_id)
+        assert invocation['state'] != 'scheduled', invocation
```

Comment:
Maybe it would be helpful to use `invocation['state'] in LIST_OF_POSSIBLE_STATES` instead, because `'failed'` would pass the assertion.

Reply:
If ``failed`` passed the assertion, the ``cancelled`` check below would fail the test. If I make it a list of ``[new, ready]`` I would need to update the list if we add more non-terminal states (one could imagine adding ``queued`` or ``waiting`` states for instance). I could reuse stuff in galaxy.models to deal with state enumerations - but I'd ultimately like the API tests to not use any of Galaxy's internals (xref https://github.com/galaxyproject/galaxy/pull/4536). So I guess I slightly prefer the current variant - but I can make the change if you still would like to see it.

Reply:
True, but:
- if the state is already `cancelled` at this point, the test would not fail
- it'd more difficult to debug the problem if the `state` here is `failed`, but the test fails only later.

Reply:
I'm not sure I agree - but I've made a bunch of changes including this and I like the tests as a whole better now.